### PR TITLE
Resolve extension members in all non-invocation contexts

### PR DESCRIPTION
**Pull request by jcouv**
When we bind a member access expression there are two cases:
1. the expression is considered "invoked" (syntactically part of an `invocation_expression`), in which case the extension member lookup will filter out non-invocable members,
2. the expression is not considered "invoked", so if member lookup failed, we can go ahead and use normal extension member lookup (ie. not filtering out non-invocable members) to resolve the expression to a non-method extension member.

This PR implements the second case.

It also removes previous logic we had for resolving to a non-method extension member during a conversion. That logic relied on an incorrect understanding of the spec (that conversions to a delegate type would cause the expression to be treated as "invoked", so we had to wait until processing the conversion to perform the extension member lookup). 

So the following scenario becomes an error:
```
var x = (System.Action)D.f; // binds to field, so error: string cannot convert to Action
System.Action a = D.f; // ditto

class D { }
 
implicit extension E1 for D
{
    public static string f = null;
}
 
implicit extension E2 for object
{
    public static void f() { }
}
```

----

Here are the relevant sections of the spec that justify this understanding:

The [Member Lookup](https://github.com/dotnet/csharpstandard/blob/draft-v8/standard/expressions.md#125-member-lookup) has concept of a member being "invoked":
> "If the simple_name or member_access occurs as the primary_expression of an invocation_expression, the member is said to be invoked."
> "Next, if the member is invoked, all non-invocable members are removed from the set."

But in [Method group conversions](https://github.com/dotnet/csharpstandard/blob/draft-v8/standard/conversions.md#108-method-group-conversions), although treat the conversion as an invocation:
> A single method M is selected corresponding to a method invocation of the form `E(A)` ... [using arguments constructed from the parameter types of delegate type]

The method group conversion only kicks in when we have already determined that we have a method group:
> "An implicit conversion exists from a method group ([§12.2](https://github.com/dotnet/csharpstandard/blob/draft-v8/standard/expressions.md#122-expression-classifications)) to a compatible delegate type ([§20.4](https://github.com/dotnet/csharpstandard/blob/draft-v8/standard/delegates.md#204-delegate-compatibility)). If D is a delegate type, and E is an expression that is classified as a method group, ..."

In short, a conversion to a delegate type does not cause the expression to be treated as "invoked", although we process it somewhat like an invocation.

Relates to test plan https://github.com/dotnet/roslyn/issues/66722

### CODE REVIEW

**Comment by jcouv** [comment]
>         // PROTOTYPE(static) the sorting mechanism for more specific extension members to hide less specific ones is insufficient and needs to be revised

:memo: I will address this in my next PR. 

The issue is that the sorting helper used in `LookupImplicitExtensionMembersInSingleBinder`  assumes transitivity.
In some orderings of the source, it only performs the following comparisons to arrive at an order: E4 vs. E3 (E4 is not more specific than E3), E1 vs. E4 (E1 is not more specific than E4), E2 vs. E1 (E2 is not more specific than E1). 
The resulting order (from more specific to less specific) that it computes is "E3, E4, E1, E2" which is incorrect, because it fails to capture the two key ordering relationships: E1 vs. E3, and E2 vs. E4.

---
Refers to: src/Compilers/CSharp/Test/Emit3/ExtensionTypeTests.cs:31130 in c4afd43. [](commit_id = c4afd43b18753386b0a5d120f91d14047258b3c2, deletion_comment = False)

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 1]
```diff
@@ -598,6 +581,13 @@ private BoundExpression BindExpressionInternal(ExpressionSyntax node, BindingDia
 
             BoundExpression result = bindExpressionInternal(node, diagnostics, invoked, indexed);
 
+            // Only invocation expressions need to resolve extension members as invoked (ie. filtering out non-invocable members)
+            // All other contexts can go ahead and resolve to a non-method extension member
+            if (!invoked)
+            {
+                result = ResolveToNonMethodExtensionMemberIfPossible(result, diagnostics);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:588]: >ResolveToNonMethodExtensionMemberIfPossible [](http://example.com/codeflow?start=25&length=43)

Is the name somewhat confusing? We may still resolve to methods, right?

**Comment by AlekseyTs** [comment]
>                 if (resolution.HasAnyErrors)

This isn't primarily related to the changes in this PR, but should we do something for extensions on this code path?

---
Refers to: src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs:1127 in bba7662. [](commit_id = bba7662055875d04191a3722e7dda2b0eefa2138, deletion_comment = False)

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs (hunk 1 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:588]: Maybe I didn't quite understand the question. `ResolveToNonMethodExtensionMemberIfPossible` only returns non-methods.

**Comment by jcouv** [comment]
>                 if (resolution.HasAnyErrors)

I think the code in `HasCollectionExpressionApplicableAddMethod` looks good as-is (we do `ResolveMethodGroup` above using `analyzedArguments`, so that'll resolve an instance method, classic extension methods and extension type methods).
But I'll add a comment to check.

Also, I already have a note in the test plan that we need to exclude extension type methods in some collection expression or params scenarios.

---
In reply to: [2085924348](https://github.com/dotnet/roslyn/pull/73239#issuecomment-2085924348) [](http://example.com/codeflow?ancestors=2085924348)

---
Refers to: src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs:1127 in bba7662. [](commit_id = bba7662055875d04191a3722e7dda2b0eefa2138, deletion_comment = False)

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 2]
```diff
@@ -7560,7 +7543,7 @@ private BoundExpression ResolveToExtensionMemberIfPossible(BoundExpression expr,
             }
         }
 
-        private BoundExpression BindInstanceMemberAccess(
+        private BoundExpression BindMemberAccessWithBoundLeftInternal(
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:7546]: >Internal [](http://example.com/codeflow?start=61&length=8)

What are we trying to convey with the "Internal" suffix? It doesn't look meaningful to me. Consider dropping it. If your goal is as simple as to not share name with the other helper, numeric suffixes "BindMemberAccessWithBoundLeft1" and "BindMemberAccessWithBoundLeft2" will be better in my opinion.

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 3]
```diff
@@ -7614,8 +7597,19 @@ private BoundExpression ResolveToExtensionMemberIfPossible(BoundExpression expr,
                         boundLeft = ReplaceTypeOrValueReceiver(boundLeft, useType: false, diagnostics);
                     }
 
-                    return MakeBoundMethodGroupAndCheckOmittedTypeArguments(boundLeft, rightName, typeArgumentsWithAnnotations, lookupResult,
+                    BoundExpression result = MakeBoundMethodGroupAndCheckOmittedTypeArguments(boundLeft, rightName, typeArgumentsWithAnnotations, lookupResult,
                         flags, node, typeArgumentsSyntax, diagnostics);
+
+                    if (!invoked)
+                    {
+                        var extensionResult = ResolveToNonMethodExtensionMemberIfPossible(result, diagnostics);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:7605]: >var extensionResult = ResolveToNonMethodExtensionMemberIfPossible(result, diagnostics); [](http://example.com/codeflow?start=24&length=87)

```result = ResolveToNonMethodExtensionMemberIfPossible(result, diagnostics);```? As we do in ```tryBindMemberAccessWithBoundTypeLeft```.

File: src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs [hunk 4]
```diff
@@ -721,14 +721,15 @@ private bool HasApplicableMemberWithPossiblyExpandedNonArrayParamsCollection<TMe
                          (analyzedArguments.HasDynamicArgument ? OverloadResolution.Options.DynamicResolution : OverloadResolution.Options.None));
             diagnostics.Add(expression, useSiteInfo);
 
-            if (resolution.IsExtensionMember(out Symbol extensionMember))
+            if (resolution.IsNonMethodExtensionMember(out Symbol extensionMember))
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs:724]: >if (resolution.IsNonMethodExtensionMember(out Symbol extensionMember)) [](http://example.com/codeflow?start=12&length=70)

Is this `if` statement still necessary? Is its body still reachable? Can you give an example of a scenario?

**Comment by AlekseyTs** [comment]
>                 if (resolution.HasAnyErrors)

I asked the question because I saw that `BindMethodGroupInvocation` does something special for non method extension members right about before similar place in code.

---
In reply to: [2086545538](https://github.com/dotnet/roslyn/pull/73239#issuecomment-2086545538) [](http://example.com/codeflow?ancestors=2086545538,2085924348)

---
Refers to: src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs:1127 in bba7662. [](commit_id = bba7662055875d04191a3722e7dda2b0eefa2138, deletion_comment = False)

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs (hunk 2 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:7546]: Yes, my main goal is to remove the `Instance` portion (misleading) and not share names since we have a cref reference to it. How about "Core" suffix?

File: src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs (hunk 4 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs:724]: Yes, we can still get here. Below is an example.
When we have an invocation, we delay the resolution of extensions when binding a member access. We resolve the extensions when binding the invocation (ie. here). So we may find that our receiver is not a method group, but instead an invocable member (for example a delegate type field). 
In such case, we bind that member access to the invocable member and invoke it with the given arguments.


```
    public void ExtensionInvocation_OnlyDelegateFieldExists()
    {
        // Invocable fields are considered during extension invocation
        var source = """
E.Field = (i) => { System.Console.Write($"ran({i})"); };
C.Field(42);

class C { }

delegate void D(int i);

implicit extension E for C
{
    public static D Field;
}
""";
```

**Comment by jcouv** [comment]
@jjonescz for second review. Thanks

**Comment by jcouv** [comment]
>                 if (resolution.HasAnyErrors)

Thanks. I identified a scenario that hits that. Blocked it off for now and left a PROTOTYPE comment

---
In reply to: [2088772543](https://github.com/dotnet/roslyn/pull/73239#issuecomment-2088772543) [](http://example.com/codeflow?ancestors=2088772543,2085924348)

---
Refers to: src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs:1127 in bba7662. [](commit_id = bba7662055875d04191a3722e7dda2b0eefa2138, deletion_comment = False)

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs (hunk 2 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: > How about "Core" suffix?

I do not find this better.  A numeric suffix is better in my opinion to avoid sharing the name. Unless we can come up with a meaningful differentiator.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: Since you already changed the name, I guess I can live with it.

**Comment by AlekseyTs** [comment]
>         // PROTOTYPE(instance) We need to decide whether to allow this and adjust HasCollectionExpressionApplicableAddMethod and other downstream logic accordingly

I am not sure what are you suggesting to allow in this scenario. Especially that the error isn't "near" a collection expression.

---
Refers to: src/Compilers/CSharp/Test/Emit3/ExtensionTypeTests.cs:37490 in f33ffce. [](commit_id = f33ffce1ad9237422e83a19f440633bbf0ba1d43, deletion_comment = False)

**Comment by AlekseyTs** [comment]
> MyCollection c = [42];

I would expect an error here, there is no suitable Add method. It must be a method, extensions do not change anything about this requirement.

---
Refers to: src/Compilers/CSharp/Test/Emit3/ExtensionTypeTests.cs:37477 in f33ffce. [](commit_id = f33ffce1ad9237422e83a19f440633bbf0ba1d43, deletion_comment = False)

File: src/Compilers/CSharp/Portable/Binder/Binder_Query.cs [hunk 5]
```diff
@@ -953,25 +955,6 @@ private static UnboundLambda MakeQueryUnboundLambda(CSharpSyntaxNode node, Query
                     // Could not find an implementation of the query pattern for source type '{0}'.  '{1}' not found.
                     diagnostics.Add(ErrorCode.ERR_QueryNoProvider, node.Location, MessageID.IDS_AnonMethod.Localize(), methodName);
                 }
-                else if (ultimateReceiver.Kind == BoundKind.MethodGroup)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Query.cs:956]: >else if (ultimateReceiver.Kind == BoundKind.MethodGroup) [](http://example.com/codeflow?start=16&length=56)

I am not sure why this code is removed. The only difference from 'main' is the PROTOTYPE comment. Which means this code wasn't touched for extension types. I think we should keep it.

**Comment by AlekseyTs** [comment]
Just an FYI, I am done reviewing compiler changes for commit 6. Planning to review tests later today.

File: src/Compilers/CSharp/Portable/Binder/Binder_Query.cs (hunk 5 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Binder/Binder_Query.cs:956]: This code is unreachable. The case above for `ultimateReceiver.HasAnyErrors` intercepts all the cases that might be handled here.
The reason I prefer to remove the code rather than just leave it is that it calls `ResolveMethodGroup` which begs the question for how extensions should be handled. But anything I do here cannot be tested.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Binder/Binder_Query.cs:956]: > This code is unreachable. The case above for ultimateReceiver.HasAnyErrors intercepts all the cases that might be handled here.

Thanks for the explanation. I am good with removal of the code.

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 6]
```diff
@@ -7466,8 +7448,15 @@ private bool IsUsingAliasInScope(string name)
                         return BindMemberOfType(node, right, rightName, rightArity, indexed, boundLeft, typeArgumentsSyntax, typeArguments, lookupResult, BoundMethodGroupFlags.SearchExtensionMethods, diagnostics: diagnostics);
                     }
 
-                    return MakeBoundMethodGroupAndCheckOmittedTypeArguments(boundLeft, rightName, typeArguments, lookupResult,
+                    BoundExpression result = MakeBoundMethodGroupAndCheckOmittedTypeArguments(boundLeft, rightName, typeArguments, lookupResult,
                         flags: BoundMethodGroupFlags.SearchExtensionMethods, node, typeArgumentsSyntax, diagnostics);
+
+                    if (!invoked)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:7454]: >if (!invoked) [](http://example.com/codeflow?start=20&length=13)

I am not sure why do we do this conditionally. Is there a specific reason to keep non-methods unresolved for invocation scenarios?

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 7]
```diff
@@ -7614,8 +7597,15 @@ private BoundExpression ResolveToExtensionMemberIfPossible(BoundExpression expr,
                         boundLeft = ReplaceTypeOrValueReceiver(boundLeft, useType: false, diagnostics);
                     }
 
-                    return MakeBoundMethodGroupAndCheckOmittedTypeArguments(boundLeft, rightName, typeArgumentsWithAnnotations, lookupResult,
+                    BoundExpression result = MakeBoundMethodGroupAndCheckOmittedTypeArguments(boundLeft, rightName, typeArgumentsWithAnnotations, lookupResult,
                         flags, node, typeArgumentsSyntax, diagnostics);
+
+                    if (!invoked)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:7603]: >if (!invoked) [](http://example.com/codeflow?start=20&length=13)

Same question

**Comment by AlekseyTs** [comment]
>     }

It feels like there are way too many tests are added given the small amount of actual code paths modified. Of course, having more tests doesn't hurt. However, consider if there is an excessive redundancy given the final state of the changes in this PR, and, perhaps, some scenarios no longer add real value.

---
Refers to: src/Compilers/CSharp/Test/Emit3/ExtensionTypeTests.cs:39614 in 0b41777. [](commit_id = 0b41777891577bf993cd19c88f13156c0a5483f3, deletion_comment = False)

**Comment by AlekseyTs** [comment]
Done with review pass (commit 8)

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs (hunk 6 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:7454]: We need arguments to resolve invocation scenarios, even if we're only interested in invocable non-method members.

Consider a classic extension method in an inner scope and an extension type with an invocable non-method member in an outer scope.
Depending on whether the extension method is applicable or not, the invocable non-method member may be found or not. 
But to determine whether it is applicable, we need arguments.

**Comment by jcouv** [comment]
>     }

Makes sense. I pruned some tests that were more relevant for a previous implementation (all scenarios calling `BindValue`)

---
In reply to: [2089257969](https://github.com/dotnet/roslyn/pull/73239#issuecomment-2089257969) [](http://example.com/codeflow?ancestors=2089257969)

---
Refers to: src/Compilers/CSharp/Test/Emit3/ExtensionTypeTests.cs:39614 in 0b41777. [](commit_id = 0b41777891577bf993cd19c88f13156c0a5483f3, deletion_comment = False)

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs (hunk 6 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:7448]: From offline discussion, it would be possible to do this unconditionally, but it would not always perform the resolution to a non-extension member (sometimes that will be possible to do only once we have arguments to discard some extension methods that are in closer scopes) and it would probably not be more readable/understandable.

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 8]
```diff
@@ -7530,37 +7515,54 @@ private BoundExpression MakeMemberAccessValue(BoundExpression expr, BindingDiagn
...
         }
 
-        private BoundExpression ResolveToExtensionMemberIfPossible(BoundExpression expr, BindingDiagnosticBag diagnostics)
+#nullable enable
+        // When we're binding a member access that is not invoked and the member lookup yielded no result:
+        // - if an extension member lookup finds a non-method extension member (closer than any classic extension method compatible with the receiver
+        //   or than any method from an extension type), then that's the member being accessed.
+        //
+        // - if the extension member lookup finds a method (classic extension method compatible with the receiver or method in extension type;
+        //   closer than any non-method extension member), we don't need to touch the result for the member access (it's a method group already).
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:7524]: >we don't need to touch the result for the member access (it's a method group already). [](http://example.com/codeflow?start=59&length=86)

It looks like this part of the comment is no longer relevant because we are no longer producing and passing in the method group.

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 9]
```diff
@@ -7530,37 +7515,54 @@ private BoundExpression MakeMemberAccessValue(BoundExpression expr, BindingDiagn
...
+        //   This method group will be resolved specially in scenarios that can handle method group
+        //   (such as inferred local `var x = A.B;`, conversion to a delegate type `System.Action a = A.B;`).
+        //   It will be an error in other scenarios.
+        //
+        // - if the extension member lookup is ambiguous, then we'll use an error symbol as the result of the member access.
+        //
+        // - if the extension member lookup finds nothing, then we don't need to touch the result for the member access.
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:7531]: >then we don't need to touch the result for the member access. [](http://example.com/codeflow?start=59&length=61)

This part as well

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 10]
```diff
@@ -7530,37 +7515,54 @@ private BoundExpression MakeMemberAccessValue(BoundExpression expr, BindingDiagn
...
-                    return expr;
-                // PROTOTYPE need to handle PropertyGroup as well
+                var memberAccess = GetExtensionMemberAccess(syntax, receiver, extensionMember,
+                    typeArgumentsSyntax, typeArgumentsOpt, diagnostics);
 
-                default:
-                    return expr;
+                if (!memberAccess.HasErrors && typeArgumentsSyntax.Any(SyntaxKind.OmittedTypeArgument))
+                {
+                    Error(diagnostics, ErrorCode.ERR_OmittedTypeArgument, syntax);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:7554]: >Error(diagnostics, ErrorCode.ERR_OmittedTypeArgument, syntax); [](http://example.com/codeflow?start=20&length=62)

Do we have a test covering this code path? Including scenario when the member is a type

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 12)

**Review by jjonescz — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs (hunk 10 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:7554]: Yes, `ExtensionMemberLookup_MatchingExtendedType_GenericType_GenericMember_OmittedTypeArgument` (a generic type scenario) and `ExtensionInvocation_TypeReceiver_TypeArguments_Omitted` and `ExtensionInvocation_InstanceReceiver_TypeArguments_Omitted` (a generic method, with either a type or an instance receiver).

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs (hunk 8 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: My intent is to add some documentation that relates to the callers (which are unfortunately split across two different methods). I'll try to clarify comment

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 13)
